File: solutions/_1239_maximum_length_of_a_concatenated_string_with_unique_characters/solution.cpp
#include <vector>
#include <string>

using namespace std;

class Solution
{
public:
    int maxLength(vector<string> &arr)
    {
        vector<int> dp = {0};
        int res = 0;
        for (const string &s : arr)
        {
            int a = 0;
            bool hasDuplicate = false;
            for (char ch : s)
            {
                int b = 1 << (ch - 'a');
                if (a & b)
                {
                    hasDuplicate = true;
                    break;
                }
                a |= b;
            }
            if (hasDuplicate)
                continue;
            for (int i = dp.size() - 1; i >= 0; --i)
            {
                bool hasOverlap = (dp[i] & a) != 0;
                if (hasOverlap)
                    continue;
                dp.push_back(dp[i] | a);
                res = max(res, __builtin_popcount(dp.back()));
            }
        }
        return res;
    }
};

/*
https://leetcode.com/problems/maximum-length-of-a-concatenated-string-with-unique-characters/
*/

File: solutions/_1239_maximum_length_of_a_concatenated_string_with_unique_characters/solution.js
/**
 * @param {string[]} arr
 * @return {number}
 */
var maxLength = function (arr) {
  let max = 0;
  const dp = [0];
  for (const str of arr) {
    let a = 0;
    let hasDuplicate = false;
    for (const ch of str) {
      const b = 1 << (ch.charCodeAt(0) - 97);
      if (a & b) {
        hasDuplicate = true;
        break;
      }
      a |= b;
    }
    if (!hasDuplicate) {
      for (let i = dp.length - 1; i >= 0; i--) {
        const hasOverlap = (dp[i] & a) !== 0;
        if (!hasOverlap) {
          const s = dp[i] | a;
          dp.push(s);
          max = Math.max(
            max,
            s
              .toString(2)
              .split("")
              .filter((bit) => bit === "1").length
          );
        }
      }
    }
  }
  return max;
};

module.exports = { maxLength };

/*
https://leetcode.com/problems/maximum-length-of-a-concatenated-string-with-unique-characters/
*/
